src/scope.rs

```diff
@@ -1,292 +1,950 @@
+        self.cast_local(|scope| {
+          raw::v8__Isolate__ThrowException(scope.get_isolate_ptr(), &*exception)
+        })
+      }
+      .unwrap()
+    }
+
+    pub(crate) unsafe fn cast_local<F, T>(
+      &mut self,
+      f: F,
+    ) -> Option<Local<'s, T>>
+    where
+      F: FnOnce(&mut Self) -> *const T,
+    {
+      // `ScopeData::get_mut()` is called here for its side effects: it checks
+      // that `self` is actually the active scope, and if necessary it will
+      // drop (escapable) handle scopes of which the drop call has been
+      // deferred.
+      data::ScopeData::get_mut(self);
+      Local::from_raw(f(self))
+    }
   }
 
-  pub(crate) fn data(&self) -> &S {
-    unsafe { &*self.data }
+  impl<'s, C> AsRef<Isolate> for HandleScope<'s, C> {
+    fn as_ref(&self) -> &Isolate {
+      self.get_isolate()
+    }
   }
 
-  pub(crate) fn data_mut(&mut self) -> &mut S {
-    unsafe { &mut *self.data }
+  impl<'s, C> AsMut<Isolate> for HandleScope<'s, C> {
+    fn as_mut(&mut self) -> &mut Isolate {
+      self.get_isolate_mut()
+    }
   }
 
-  pub(crate) fn parent(&self) -> &P {
-    &self.parent
+  unsafe impl<'s, C> Scope for HandleScope<'s, C> {}
+
+  impl<'s> Deref for HandleScope<'s, ()> {
+    type Target = Isolate;
+    fn deref(&self) -> &Self::Target {
+      self.get_isolate()
+    }
   }
 
-  pub(crate) fn parent_mut(&mut self) -> &mut P {
-    &mut self.parent
+  impl<'s> DerefMut for HandleScope<'s, ()> {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+      self.get_isolate_mut()
+    }
   }
-}
 
-/// A CallbackScope can be used to obtain a mutable Isolate reference within
-/// a callback that is called by V8 on the thread that already has a Locker
-/// on the stack.
-///
-/// Using a CallbackScope in any other situation is unsafe.
-/// Also note that CallbackScope should not be used for function and property
-/// accessor callbacks; use FunctionCallbackScope and PropertyCallbackScope
-/// instead.
-///
-/// A CallbackScope can be created from the following inputs:
-///   - `Local<Context>`
-///   - `Local<Message>`
-///   - `Local<Object>`
-///   - `Local<Promise>`
-///   - `Local<SharedArrayBuffer>`
-///   - `&PromiseRejectMessage`
-pub struct CallbackScope<X = Contained> {
-  isolate: *mut Isolate,
-  phantom: PhantomData<X>,
-}
+  impl<'s> Deref for HandleScope<'s> {
+    type Target = ContextScope<'s>;
+    fn deref(&self) -> &Self::Target {
+      self.cast_ref()
+    }
+  }
 
-pub struct Contained;
-pub struct Escapable;
+  impl<'s> DerefMut for HandleScope<'s> {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+      self.cast_mut()
+    }
+  }
 
-impl<'s> CallbackScope {
-  pub fn new<I>(input: I) -> Scope<'s, Self>
+  impl<'s, C> Drop for HandleScope<'s, C> {
+    fn drop(&mut self) {
+      data::ScopeData::get_mut(self).notify_scope_dropped();
+    }
+  }
+
+  pub unsafe trait NewHandleScopeParam<'s> {
+    type NewScope: Scope;
+    fn get_isolate_mut(&mut self) -> &mut Isolate;
+  }
```

Comment:
Where is `NewScope` used?

Reply:
https://github.com/denoland/rusty_v8/pull/406/files#diff-7a8f2fd4af9cbf971ac52735898b4eacR186

Reply:
https://github.com/denoland/rusty_v8/blob/af0738c595ef0d3bd5a1a1d333b81bc5138d5d57/src/scope.rs#L186-L193